lib/scim/sdk/client.go

```diff
@@ -0,0 +1,643 @@
+			Operations: []PatchOp{
+				{
+					Operation: patchOp,
+					Path:      "members",
+					Value:     membersPage,
+				},
+			},
+		}
+		payload, err := json.Marshal(res)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		resp, err := c.do(ctx, u, http.MethodPatch, bytes.NewReader(payload))
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		// Close the response body immediately after the request is done..
+		resp.Body.Close()
+		switch resp.StatusCode {
+		case http.StatusOK, http.StatusNoContent:
+		default:
+			return decodeError(resp)
+		}
+		// See https://docs.aws.amazon.com/singlesignon/latest/developerguide/patchgroup.html
+		// * A maximum of 100 membership changes are allowed in a single request.
+		// Bypass this limitation by appending the next 100 members to the group.
+		patchOp = OpAdd
+	}
+	return nil
+}
+
+func (c *client) endpointURL(paths ...string) (*url.URL, error) {
+	path, err := url.JoinPath(c.Endpoint, paths...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	u, err := url.Parse(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return u, nil
+}
+
+// ListGroups returns a list of groups from the SCIM server.
+// NOTE: Some implementations may not support pagination or filtering.
+// AWS IC SCIM Limitation see https://docs.aws.amazon.com/singlesignon/latest/developerguide/listgroups.html
+//   - ListGroups return an empty member list.
+//   - At this time, the ListGroups API is only capable of returning up to 50 results.
+func (c *client) ListGroups(ctx context.Context, queryOptions ...QueryOption) (*ListGroupResponse, error) {
+	var options QueryOptions
+	for _, opt := range queryOptions {
+		opt(&options)
+	}
+
+	u, err := c.endpointURL("Groups")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	u.RawQuery = options.toQuery().Encode()
+	u.RawQuery = u.Query().Encode()
+
+	resp, err := c.do(ctx, u, http.MethodGet, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+	default:
+		return nil, decodeError(resp)
+	}
+
+	var listResp ListGroupResponse
+	if err := json.NewDecoder(resp.Body).Decode(&listResp); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &listResp, nil
+}
+
+// Ping checks the connection to the SCIM server by sending a request to the ServiceProviderConfig endpoint.
+func (c *client) Ping(ctx context.Context) error {
+	u, err := c.endpointURL("ServiceProviderConfig")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	resp, err := c.do(ctx, u, http.MethodGet, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	switch resp.StatusCode {
+	case http.StatusOK:
+	case http.StatusUnauthorized, http.StatusForbidden:
+		return trace.AccessDenied("unauthorized")
+	case http.StatusInternalServerError:
+		return trace.BadParameter("internal server error")
+	default:
+		return trace.BadParameter("unexpected status code %v", resp.StatusCode)
+	}
+	return nil
+}
```

Comment:
Instead of moving the whole scim sdk to oss, can't we just moved this logic to oss? Convert this to a function that accepts a ready made http client param and we have a common validator that can be used in both tctl and web ui?

Reply:
Seems reasonable, done.